File: baekjoon/python/gold3/q21606_morning_walk.py
# https://www.acmicpc.net/problem/21606
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

N = int(input().rstrip())

graph = [[] for _ in range(N+1)]


def success_from_other():
    in_out = '0' + input().rstrip()
    for _ in range(N-1):
        x, y = map(int, input().rstrip().split())
        graph[x].append(y)
        graph[y].append(x)


    def dfs(start):
        visited[start] = 1
        ret = 0
        for des in graph[start]:
            if not visited[des]:
                if in_out[des] == '1':
                    ret += 1
                else:
                    ret += dfs(des)
        return ret

    ans = 0    
    visited = [0] * (N+1)
    for i in range(1, N+1):
        c = in_out[i]
        if c == '1':
            for des in graph[i]:
                if in_out[des] == '1':
                    ans += 1
        else:
            if not visited[i]:
                ret = dfs(i)
                ans += ret * (ret-1)
    print(ans)
success_from_other()

def fail():
    in_out = [0] * (N+1) ## 1이 실내 
    in_points = []
    for i, c in enumerate(input().rstrip()):
        c = int(c)
        in_out[i+1] = c
        if c == 1: 
            in_points.append(i+1)
    for _ in range(N-1):
        x, y = map(int, input().rstrip().split())
        graph[x].append(y)
        graph[y].append(x)

    def sol(start, cnt):
        visited[start] = 1
        for des in graph[start]:
            if not visited[des]:
                if in_out[des]:
                    visited[des] = 1
                    cnt += 1
                else:
                    cnt = sol(des, cnt)
        return cnt

    ans = 0
    for start in in_points:
        visited = [0] * (N+1)
        ans += sol(start, 0)

    print(ans)

# fail()

File: baekjoon/java/src/gold3/Q21606_MorningWalk.java
// https://www.acmicpc.net/problem/21606
package gold3;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Q21606_MorningWalk {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static void input() {
        N = scan.nextInt();
        inOut = new int[N+1];
        String next = scan.next();
        for (int i = 0; i < N; i++) {
            if(next.charAt(i) == '1') {
                inOut[i+1] = 1;
            }
        }
        graph = new ArrayList<>();
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < N-1; i++) {
            int a = scan.nextInt();
            int b = scan.nextInt();
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        visited = new boolean[N+1];
    }

    static int N;
    static int[] inOut;
    static List<List<Integer>> graph;
    static boolean[] visited;

    static void sol() {
        int ans = 0;
        for (int i = 1; i <= N; i++) {
            if(inOut[i] == 1) {
                for(int des:graph.get(i)) {
                    if(inOut[des] == 1) {
                        ans++;
                    }
                }
            } else {
                if(!visited[i]) {
                    int reuslt = dfs(i);
                    ans += reuslt * (reuslt - 1);
                }
            }
        }
        System.out.println(ans);
    }

    static int dfs(int start) {
        visited[start] = true;
        int ret = 0;
        for(int des: graph.get(start)) {
            if(!visited[des]) {
                if(inOut[des] == 1) {
                    ret++;
                } else {
                    ret += dfs(des);
                }
            }
        }
        return ret;
    }

    public static void main(String[] args) {
        input();
        sol();
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}
